### PR TITLE
[Task-4] As a user, I can register

### PR DESCRIPTION
**Pull request by mcneto**
Task - [[Task-4] As a user, I can register](https://github.com/mcneto/ruby-google-search-ic/issues/7)

## What happened

- Implemented User Authentication using Devise
- User can register

## Insight

- User model and most of this part was auto generated by Devise
- Add custom attributes (First Name, Last Name) to exiting model and other configurations
- User can register if doesn't have an account already
- Move testing part to a new task (#40)

## Proof Of Work
**Register**
Screen Recording 2564-06-28 at 12 17 24
**Error Message**
Screen Shot 2564-06-28 at 11 50 06

### CODE REVIEW

**Comment by junan** [comment]
@mcneto As this PR is a backend PR, so title should be `[Task-4] [Backend] As a user, I can register`

**Comment by junan** [comment]
@mcneto The testing should be added in the same PR, not in a separate PR

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
+  validates :first_name, :last_name, presence: true
```
Comment (junan) [inline app/models/user.rb:8]:
```suggestion

  validates :first_name, :last_name, presence: true
```

File: db/migrate/20210621044346_devise_create_users.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
+# frozen_string_literal: true
+
+class DeviseCreateUsers < ActiveRecord::Migration[6.1]
+  def change
+    create_table :users do |t|
+      ## Database authenticatable
+      t.string :email,              null: false, default: ""
```
Comment (junan) [inline db/migrate/20210621044346_devise_create_users.rb:7]: Based on our convention, the email field type should be postgres `citext` type. You can check Xavier repo for reference 
Screen Shot 2021-06-29 at 12 37 38 PM

**Comment by mcneto** [comment]
> @mcneto As this PR is a backend PR, so title should be `[Task-4] [Backend] As a user, I can register`

I want to use a label for it. Can I use both or its prefers to be displayed only on the title?

File: app/views/devise/unlocks/new.html.erb [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+<%= form_for(resource, as: resource_name, url: unlock_path(resource_name), html: { method: :post }) do |f| %>
+  <%= render "devise/shared/error_messages", resource: resource %>
+
+  <div class="field">
+    <%= f.label :email %><br />
+    <%= f.email_field :email, autofocus: true, autocomplete: "email" %>
+  </div>
+
+  <div class="actions">
```
Comment (junan) [inline app/views/devise/unlocks/new.html.erb:11]: based on out css convention the class name should be singular and need to use two word
```suggestion
  <div class="user-action">
```

**Comment by mcneto** [comment]
> @mcneto The testing should be added in the same PR, not in a separate PR

@junan I separate it because I took so much time in the test part that may cause to not have any feature update ready for a release or review but I'm currently working on it I wait to add it to this task then.

**Comment by junan** [comment]
> > @mcneto The testing should be added in the same PR, not in a separate PR

> @junan I separate it because I took so much time in the test part that may cause to not have any feature update ready for a > release or review, but I'm currently working on it I wait to add it to this task then.

@mcneto I understood that sometimes it takes much time to learn/work with something new. But as per out convention, it is important to add test to the same PR. So ya please add test in the same PR. Have a look Xavier PR for reference.

**Comment by junan** [comment]
> @junan I want to use a label for it. Can I use both or its prefers to be displayed only on the title?

Thats a good question! Some developer uses the type in the PR title also and just checked we have not defined anything in the conventions https://nimblehq.co/compass/development/code-reviews/

@mcneto Can you create an RFC issue for it in the compass some the team member vote and finalize the convention. 

Like,
1) Should we use the back-end / front-end label  in the PR  title along with github label
2) Should we use the back-end / front-end label on the in the Github label only


You can check this example RRC https://github.com/nimblehq/compass/issues/642

File: app/models/user.rb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+class User < ApplicationRecord
+  # Include default devise modules. Others available are:
```
Comment (rafayet-monon) [inline app/models/user.rb:4]: Let's git rid of the commented lines if not necessary.

File: app/views/devise/shared/_links.html.erb [hunk 5]
```diff
@@ -0,0 +1,25 @@
+<%- if controller_name != 'sessions' %>
+  <%= link_to "Log in", new_session_path(resource_name) %><br />
+<% end %>
+
+<%- if devise_mapping.registerable? && controller_name != 'registrations' %>
+  <%= link_to "Sign up", new_registration_path(resource_name) %><br />
```
Comment (rafayet-monon) [inline app/views/devise/shared/_links.html.erb:6]: Let's use single-quote inside `erb` tags and make it consistent along the codebase.

File: app/views/devise/registrations/edit.html.erb [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+<%= form_for(resource, as: resource_name, url: registration_path(resource_name), html: { method: :put }) do |f| %>
+  <%= render "devise/shared/error_messages", resource: resource %>
+
+  <div class="field">
+    <%= f.label :"First Name" %><br />
+    <%= f.text_field :first_name, autofocus: true, autocomplete: "given-name" %>
+  </div>
+  
+  <div class="field">
+    <%= f.label :"Last Name" %><br />
```
Comment (rafayet-monon) [inline app/views/devise/registrations/edit.html.erb:12]: I think it's necessary to use `:` if you are using quotes. 🤔

**Review by rafayet-monon — COMMENTED** [review]
There's a lot of changes in this PR that are not related to user registration. Please have a look at this comment - https://github.com/mcneto/ruby-google-search-ic/pull/39#pullrequestreview-694799211

File: app/views/devise/passwords/edit.html.erb [hunk 7]
```diff
@@ -0,0 +1,25 @@
+<h2>Change your password</h2>
+
+<%= form_for(resource, as: resource_name, url: password_path(resource_name), html: { method: :put }) do |f| %>
+  <%= render "devise/shared/error_messages", resource: resource %>
```
Comment (rafayet-monon) [inline app/views/devise/passwords/edit.html.erb:4]: @mcneto Let's use `single-quotes` in all the `erb` tags. And let's make it consistent across the codebase 🙌

File: app/views/devise/shared/_error_messages.html.erb [hunk 8]
```diff
@@ -0,0 +1,15 @@
+<% if resource.errors.any? %>
+  <div id="error_explanation">
+    <h2>
+      <%= I18n.t('errors.messages.not_saved',
```
Comment (junan) [inline app/views/devise/shared/_error_messages.html.erb:4]:
```suggestion
      <%= t('errors.messages.not_saved',
```

File: app/models/user.rb [hunk 9]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+class User < ApplicationRecord
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
+
+  validates :first_name, :last_name, presence: true
```
Comment (junan) [inline app/models/user.rb:7]: Lets add a spec for it in `spec/models/user_spec.rb`
```ruby
  describe 'Validations' do
    it { is_expected.to validate_presence_of(:first_name) }
    it { is_expected.to validate_presence_of(:last_name) }
  end
  ```
 [Reference code](https://github.com/nimblehq/the1-web/blob/f97fcb48e50e605a92b35398dbd8e3ec6324a990/spec/models/the1_contact/partner_spec.rb#L7)

File: app/views/devise/shared/_error_messages.html.erb (hunk 8 above)
Comment (mcneto) [reply app/views/devise/shared/_error_messages.html.erb]: Resolved in 4320a6c

File: app/models/user.rb (hunk 9 above)
Comment (mcneto) [reply app/models/user.rb:7]: Added in b236269

File: Gemfile [hunk 10]
```diff
@@ -84,6 +84,8 @@ group :test do
   gem 'vcr' # Gem for recording test suite's HTTP interactions
   gem 'timecop' # Gem for time travel
   gem 'rails-controller-testing' # Gem that allow to use assigns as well ass assert_template
+  gem 'factory_bot_rails' # Allow to create an object or a collection of objects with predefined sets of values for tests
+  gem 'faker'
```
Comment (junan) [inline Gemfile:88]: lets add a description
```suggestion
  gem 'faker', # A library for generating fake data
```

File: spec/systems/registration_spec.rb [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+require 'rails_helper'
+
+describe 'Registration', type: :system do
+  context 'given the valid inputs' do
+    it 'displays successful register message' do
+      sign_up_user 'test@example.com', 'test1234', 'test1234'
+      expect(page).to have_content 'Welcome! You have signed up successfully.'
+    end
+
+    it 'create new user' do
```
Comment (junan) [inline spec/systems/registration_spec.rb:12]:
```suggestion
    it 'creates a new user' do
```

File: spec/systems/registration_spec.rb [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+      sign_up_user 'test@example.com', 'test1234', 'test1234'
+      expect(User.count).to eq(1)
+    end
+  end
+
+  context 'given no inputs' do
+    it 'does not create new user' do
```
Comment (junan) [inline spec/systems/registration_spec.rb:19]:
```suggestion
    it 'does not create a new user' do
```

**Review by junan — APPROVED** [review]
I approved with a few minor comments. Good work 👍

File: Gemfile (hunk 10 above)
Comment (mcneto) [reply Gemfile]: Added in 71ba7a0

File: spec/systems/registration_spec.rb (hunk 12 above)
Comment (mcneto) [reply spec/systems/registration_spec.rb]: Resolved in 3ee5f0c

File: spec/systems/registration_spec.rb (hunk 11 above)
Comment (mcneto) [reply spec/systems/registration_spec.rb]: Resolved in 3ee5f0c

**Review by rafayet-monon — APPROVED** [review]

**Pull request merged**